### PR TITLE
ENH: Add prod to masked_reductions

### PR DESCRIPTION
**Pull request by dsaxton**
Adding prod to /core/array_algos/masked_reductions.py and using them for IntegerArray and BooleanArray. This also seems to offer a decent speedup over nanops like the other reductions:

```python
# Branch

[ins] In [3]: %timeit arr.prod()  # arr = pd.Series([None, 0, 1, 2] * 10_000, dtype="Int64")
102 µs ± 379 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

[ins] In [5]: %timeit arr.prod()  # arr = pd.Series([0, 0, 1, 2] * 10_000, dtype="Int64")
82.6 µs ± 752 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

# Master

[ins] In [4]: %timeit arr.prod()  # arr = pd.Series([None, 0, 1, 2] * 10_000, dtype="Int64")
291 µs ± 6.23 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

[ins] In [6]: %timeit arr.prod()  # arr = pd.Series([0, 0, 1, 2] * 10_000, dtype="Int64")
78.6 µs ± 2.5 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

### CODE REVIEW

File: pandas/core/array_algos/masked_reductions.py [hunk 1]
```diff
@@ -63,17 +68,13 @@ def _minmax(func, values: np.ndarray, mask: np.ndarray, skipna: bool = True):
...
-            if values.size:
-                return func(values)
-            else:
-                # min/max with empty array raise in numpy, pandas returns NA
-                return libmissing.NA
+            return func(values)
     else:
-        subset = values[~mask]
-        if subset.size:
+        if not mask.all():
```
Comment (jorisvandenbossche) [inline pandas/core/array_algos/masked_reductions.py:77]: Doing a `mask.all()` might on average be expensive here, since in most cases you will actually need the subset afterwards

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks!
Added a small comment

**Comment by jorisvandenbossche** [comment]
Can you also add prod to the existing whatsnew note about this?

File: pandas/core/array_algos/masked_reductions.py (hunk 1 above)
Comment (dsaxton) [reply pandas/core/array_algos/masked_reductions.py:77]: Ah yes, I see what you're saying. I'll switch that back

**Comment by dsaxton** [comment]
Hmm, looks like we are hitting an overflow issue in /tests/extension/base/reduce.test_reduce_series. This is interesting because I think the current implementation is only passing this test "by accident" since the input gets cast to float64 whenever it contains NA values (without this casting we would still get an overflow): https://github.com/pandas-dev/pandas/blob/master/pandas/core/arrays/integer.py#L571

Is this test actually correct (i.e., if an integer input would overflow, do we "expect" the output we would get if it was first cast as float)? https://github.com/pandas-dev/pandas/blob/master/pandas/tests/extension/base/reduce.py#L19

**Comment by jbrockmendel** [comment]
@dsaxton shot in the dark maybe algorithms.checked_add_with_arr has something useful?

**Comment by jorisvandenbossche** [comment]
I suppose this is just the test that was not correct. Also for int64 dtype (default numpy one), we let it silently overflow instead of casting to float:

```python
In [1]: a = np.arange(1, 101)  

In [2]: pd.Series(a).prod()   
Out[2]: 0

# no missing values -> use int algo from numpy
In [3]: pd.Series(a, dtype="Int64").prod()     
Out[3]: 0

# on this branch, also this returns 0
In [4]: pd.Series(list(a) + [None], dtype="Int64").prod()     
Out[4]: 93326215443944102188325606108575267240944254854960571509166910400407995064242937148632694030450512898042989296944474898258737204311236641477561877016501813248
```

So certainly having it differ depending on the presence of missing values or not as the current implementation gives is clearly a bug.

File: doc/source/whatsnew/v1.1.0.rst [hunk 2]
```diff
@@ -356,7 +356,7 @@ Performance improvements
   sparse values from ``scipy.sparse`` matrices using the
   :meth:`DataFrame.sparse.from_spmatrix` constructor (:issue:`32821`,
   :issue:`32825`,  :issue:`32826`, :issue:`32856`, :issue:`32858`).
-- Performance improvement in reductions (sum, min, max) for nullable (integer and boolean) dtypes (:issue:`30982`, :issue:`33261`).
+- Performance improvement in reductions (sum, prod, min, max) for nullable (integer and boolean) dtypes (:issue:`30982`, :issue:`33261`).
```
Comment (jorisvandenbossche) [inline doc/source/whatsnew/v1.1.0.rst:359]: Can you add the issue number of this PR here as well?

File: pandas/core/array_algos/masked_reductions.py [hunk 3]
```diff
@@ -86,3 +88,15 @@ def min(values: np.ndarray, mask: np.ndarray, skipna: bool = True):
 
 def max(values: np.ndarray, mask: np.ndarray, skipna: bool = True):
     return _minmax(np.max, values=values, mask=mask, skipna=skipna)
+
+
+def sum(values: np.ndarray, mask: np.ndarray, skipna: bool = True, min_count: int = 0):
```
Comment (jorisvandenbossche) [inline pandas/core/array_algos/masked_reductions.py:93]: Maybe put those just under `_sumprod` to keep the helper and final functions close

File: pandas/tests/extension/test_integer.py [hunk 4]
```diff
@@ -238,9 +238,12 @@ def check_reduce(self, s, op_name, skipna):
         # overwrite to ensure pd.NA is tested instead of np.nan
         # https://github.com/pandas-dev/pandas/issues/30958
         result = getattr(s, op_name)(skipna=skipna)
-        expected = getattr(s.astype("float64"), op_name)(skipna=skipna)
-        if np.isnan(expected):
+        if skipna:
+            expected = getattr(s.dropna().astype("int64"), op_name)(skipna=skipna)
```
Comment (jorisvandenbossche) [inline pandas/tests/extension/test_integer.py:242]: +1, that seems better ;)

Comment (jorisvandenbossche) [reply pandas/tests/extension/test_integer.py:242]: But actually, if there are no NAs, the dropna of this line is also fine. So I think you can do:

if not skipna and s.isna().any():
    expected = pd.NA
else:
    expected = ... (with s.dropn().astype(int64) ..

File: pandas/core/arrays/boolean.py [hunk 5]
```diff
@@ -698,7 +698,15 @@ def _reduce(self, name: str, skipna: bool = True, **kwargs):
 
         if name in {"sum", "prod", "min", "max"}:
             op = getattr(masked_reductions, name)
-            return op(data, mask, skipna=skipna, **kwargs)
+            result = op(data, mask, skipna=skipna, **kwargs)
+
+            # if we have numeric op that would result in an int, coerce to int if possible
+            if name == "prod" and notna(result):
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/boolean.py:704]: why did you need to add this back?

Comment (dsaxton) [reply pandas/core/arrays/boolean.py:704]: I was getting a failure on some builds for \tests\arrays\boolean\test_reduction.py where we're checking that the product has int dtype so it might be needed:
```
        elif op == "prod":
>           assert isinstance(getattr(s, op)(), np.int64)
E           AssertionError: assert False
```

File: pandas/tests/extension/test_integer.py (hunk 4 above)
Comment (dsaxton) [reply pandas/tests/extension/test_integer.py:242]: Nice 👍

File: pandas/core/arrays/boolean.py [hunk 6]
```diff
@@ -696,9 +696,18 @@ def _reduce(self, name: str, skipna: bool = True, **kwargs):
         data = self._data
         mask = self._mask
 
-        if name in {"sum", "min", "max"}:
+        if name in {"sum", "prod", "min", "max"}:
             op = getattr(masked_reductions, name)
-            return op(data, mask, skipna=skipna, **kwargs)
+            result = op(data, mask, skipna=skipna, **kwargs)
+
```
Comment (jreback) [inline pandas/core/arrays/boolean.py:702]: can you try using maybe_cast_result_dtype here (you will have to add the prod operation in side that)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/arrays/boolean.py (hunk 5 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/boolean.py:705]: Hmm, then maybe the test needs to be edited. In any case, we should investigate why it is failing / what we should be expecting. 

From a quick test, it seems that numpy returns int64:

```
In [6]: np.array([], dtype="bool").prod()                                                                                                                                                                          
Out[6]: 1

In [7]: type(_)                                                                                                                                                                                                    
Out[7]: numpy.int64

In [8]: np.array([True, False], dtype="bool").prod()                                                                                                                                                               
Out[8]: 0

In [9]: type(_)                                                                                                                                                                                                    
Out[9]: numpy.int64
```

So I think we should follow that behaviour here

File: pandas/core/arrays/boolean.py (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/boolean.py:702]: As far as I understand, that should not be needed. Numpy should give the desired result for booleans.

File: pandas/core/arrays/boolean.py (hunk 5 above)
Comment (dsaxton) [reply pandas/core/arrays/boolean.py]: From what I can tell it was something that was only affecting Windows builds: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=32986&view=logs&jobId=077026cf-93c0-54aa-45e0-9996ba75f6f7&j=077026cf-93c0-54aa-45e0-9996ba75f6f7&t=e95cf409-86ae-5b4d-6c5f-79395ef75e8f

File: pandas/core/arrays/boolean.py [hunk 7]
```diff
@@ -686,9 +686,13 @@ def _reduce(self, name: str, skipna: bool = True, **kwargs):
...
         mask = self._mask
 
-        if name in {"sum", "min", "max"}:
+        if name in {"sum", "prod", "min", "max"}:
             op = getattr(masked_reductions, name)
-            return op(data, mask, skipna=skipna, **kwargs)
+            result = op(data, mask, skipna=skipna, **kwargs)
+            dtype = maybe_cast_result_dtype(dtype=data.dtype, how=name)
+            if notna(result) and (dtype != result.dtype):
+                result = result.astype(dtype)
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/boolean.py:694]: I am still not sure we actually need to do this. We could also choose to follow numpy's behaviour to return platform int (any idea what we do for "bool" dtype?)

BTW, this should also change the result for `sum`, so this was not tested?

Comment (dsaxton) [reply pandas/core/arrays/boolean.py]: @jorisvandenbossche I think you're right and it was actually the test that needed to change here (np.int64 -> np.int_)

Comment (jorisvandenbossche) [reply pandas/core/arrays/boolean.py]: Yeah, in any case that's what I did for sum, I see now (so if we decide on following numpy vs always returning int64, we should do it for both sum and prod)

Comment (jorisvandenbossche) [reply pandas/core/arrays/boolean.py]: (but I am fine with following numpy)

**Review by jorisvandenbossche — APPROVED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @dsaxton
